### PR TITLE
333 - Logical XOR

### PR DESCRIPTION
**Pull request by ApexTheory**
To review at your convenience.

### CODE REVIEW

File: ergotree-interpreter/src/eval/bin_op.rs [hunk 1]
```diff
@@ -591,10 +609,10 @@ mod tests {
         }
 
         #[test]
-        fn test_and_or(l in any::<bool>(), r in any::<bool>()) {
-            prop_assert_eq!(eval_relation_op(RelationOp::And, l, r), l && r);
-            prop_assert_eq!(eval_relation_op(RelationOp::Or, l, r), l || r);
+        fn test_and_or_xor(l in any::<bool>(), r in any::<bool>()) {
+            prop_assert_eq!(eval_logical_op(LogicalOp::And, l, r), l && r);
+            prop_assert_eq!(eval_logical_op(LogicalOp::Or, l, r), l || r);
+            prop_assert_eq!(eval_logical_op(LogicalOp::Xor, l, r), (l || r) && !(l && r));
```
Comment (greenhat) [inline ergotree-interpreter/src/eval/bin_op.rs:615]:
```suggestion
            prop_assert_eq!(eval_logical_op(LogicalOp::Xor, l, r), l ^ r);
```
Rust has logical XOR operator - https://doc.rust-lang.org/reference/types/boolean.html#logical-xor. Please, use it in the eval itself as well.

**Review by greenhat — CHANGES_REQUESTED** [review]
Looking great! Thank you! See one minor comment.

File: ergotree-interpreter/src/eval/bin_op.rs (hunk 1 above)
Comment (ApexTheory) [reply ergotree-interpreter/src/eval/bin_op.rs]: Of course, I should have tested that!  Fixed.

File: ergotree-interpreter/src/eval/bin_op.rs [hunk 2]
```diff
@@ -174,26 +174,33 @@ impl Evaluable for BinOp {
...
                 } else {
                     false
                 })),
-                RelationOp::Or => Ok(Value::Boolean(if !lv.try_extract_into::<bool>()? {
+                LogicalOp::Or => Ok(Value::Boolean(if !lv.try_extract_into::<bool>()? {
                     rv()?.try_extract_into::<bool>()?
                 } else {
                     true
                 })),
+                LogicalOp::Xor => Ok(Value::Boolean(if lv.try_extract_into::<bool>()? {
```
Comment (greenhat) [inline ergotree-interpreter/src/eval/bin_op.rs:191]: I meant to use rust ^ XOR in op eval as well.

Comment (ApexTheory) [reply ergotree-interpreter/src/eval/bin_op.rs]: Fixed

**Review by greenhat — APPROVED** [review]
Great! Thank you!

**Pull request merged**